integrations/jta/jdbc/src/main/java/io/helidon/integrations/jta/jdbc/LocalXAResource.java

```diff
@@ -0,0 +1,1031 @@
+                    sqlException = s;
+                } else {
+                    sqlException = null;
+                }
+                if (sqlException != null) {
+                    String sqlState = sqlException.getSQLState();
+                    if (sqlState != null
+                        && (sqlState.startsWith("080")
+                            || sqlState.equalsIgnoreCase("08S01") // ("ess" not "five")
+                            || sqlState.equalsIgnoreCase("JZ006"))) {
+                        // Connection-related database error; might be transient; use XAER_RMFAIL instead of XAER_RMERR,
+                        // apparently.  See, for example,
+                        // https://github.com/pgjdbc/pgjdbc/commit/e5aab1cd3e49051f46298d8f1fd9f66af1731299. Also see
+                        // https://github.com/pgjdbc/pgjdbc/blob/98c04a0c903e90f2d5d10a09baf1f753747b2556/pgjdbc/src/main/java/org/postgresql/xa/PGXAConnection.java#L651-L657
+                        // and
+                        // https://github.com/pgjdbc/pgjdbc/blob/98c04a0c903e90f2d5d10a09baf1f753747b2556/pgjdbc/src/main/java/org/postgresql/xa/PGXAConnection.java#L553. Also
+                        // see
+                        // https://github.com/ironjacamar/ironjacamar/blob/ff62b8b23f59f9fbb9c15be40fef38efb872c436/core/src/main/java/org/jboss/jca/core/tx/jbossts/LocalConnectableXAResourceImpl.java#L55-L61.
+                        //
+                        // But also note XAER_RMERR vs. XAER_RMFAIL changes semantics depending on the routine (start, end,
+                        // commit, rollback, prepare, forget, recover).
+                        return (XAException) new XAException(XAER_RMFAIL).initCause(e);
+                    }
+                }
+            }
+        }
+        return (XAException) new XAException(XAER_RMERR).initCause(e);
+    }
+
+    // (Invoked only in context of a remapping BiFunction, from computeAssociation().)
+    private static Association remap(Xid xid,
+                                     Association a,
+                                     EnumSet<Association.BranchState> legalBranchStates,
+                                     UnaryOperator<Association> remapOperator) {
+        if (a == null) {
+            throw new UncheckedXAException((XAException) new XAException(XAER_NOTA)
+                                           .initCause(new NullPointerException("xid: " + xid + "; association: null")));
+        } else if (!legalBranchStates.contains(a.branchState())) {
+            throw new UncheckedXAException((XAException) new XAException(XAER_PROTO)
+                                           .initCause(new IllegalStateException("xid: " + xid + "; association: " + a)));
+        }
+        return remapOperator.apply(a);
+    }
+
+    // (Remapping BiFunction. Used in end() above.)
+    private static Association activeToIdle(Xid x, Association a) {
```

Comment:
For reviewers reading sequentially: you'll find all this "active" and "idle" talk and so on confusing unless you scroll to the bottom of this file and read a bit about the arcane branch association states defined by the specification.